src/Compilers/CSharp/Portable/Compilation/CSharpCompilation.cs

```diff
@@ -680,23 +686,152 @@ public new CSharpCompilation AddSyntaxTrees(IEnumerable<SyntaxTree> trees)
+            seenFiles.Add(tree.FilePath);
+
+            var resolver = options.SourceReferenceResolver;
+            ArrayBuilder<LoadDirective> loadDirectives = null;
+            foreach (var directive in tree.GetCompilationUnitRoot().GetLoadDirectives())
+            {
+                SyntaxTree loadedTree = null;
+                var diagnostics = DiagnosticBag.GetInstance();
+
+                var fileToken = directive.File;
+                var path = (string)fileToken.Value;
+                if (path == null)
+                {
+                    // If there is no path, the parser should have some Diagnostics to report.
+                    Debug.Assert(tree.GetDiagnostics().Any());
+                    continue;
+                }
+
+                var resolvedFilePath = resolver.ResolveReference(path, baseFilePath: null);
+                if (resolvedFilePath == null)
+                {
+                    diagnostics.Add(
+                        messageProvider.CreateDiagnostic(
+                            (int)ErrorCode.ERR_NoSourceFile,
+                            fileToken.GetLocation(),
+                            path,
+                            CSharpResources.CouldNotFindFile));
+                }
+                else if (!seenFiles.Contains(resolvedFilePath))
                 {
-                    throw new ArgumentException(CSharpResources.SubmissionCanHaveAtMostOne, "trees");
+                    var readTextDiagnostics = new List<Diagnostic>();
+                    var code = resolver.ReadText(resolvedFilePath, fileToken.GetLocation(), readTextDiagnostics);
+                    if (code == null)
+                    {
+                        Debug.Assert(readTextDiagnostics.Any());
+                        diagnostics.AddRange(readTextDiagnostics);
+                    }
+                    else
+                    {
+                        loadedTree = SyntaxFactory.ParseSyntaxTree(code, parseOptions, resolvedFilePath);
+
+                        AppendSyntaxTreesHelper(
+                            treesBuilder,
+                            loadedTree,
+                            parseOptions,
+                            options,
+                            seenFiles,
+                            messageProvider,
+                            isSubmission,
+                            ordinalMapBuilder,
+                            ref loadDirectiveMap,
+                            ref declMap,
+                            ref declTable,
+                            ref referenceDirectivesChanged);
+                    }
+                }
+                else
+                {
+                    // The path resolved, but we've seen this file before,
+                    // so don't attempt to load it again.
+                    Debug.Assert(diagnostics.IsEmptyWithoutResolution);
+                    diagnostics.Free();
+                    continue;
```

Comment:
+1 on reference counting.  Since only the first directive that loads a given tree gets a map entry, we have no way of knowing whether another tree would have added it later.